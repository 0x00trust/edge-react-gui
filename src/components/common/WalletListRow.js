// @flow

import * as React from 'react'
import { Image, StyleSheet, Text, TouchableHighlight, TouchableWithoutFeedback, View } from 'react-native'
import { Actions } from 'react-native-router-flux'
import { connect } from 'react-redux'

import { selectWallet } from '../../actions/WalletActions.js'
import { getSpecialCurrencyInfo, TRANSACTION_LIST, WALLET_LIST_SCENE } from '../../constants/indexConstants.js'
import { WALLET_LIST_OPTIONS_ICON } from '../../constants/WalletAndCurrencyConstants.js'
import s from '../../locales/strings.js'
import T from '../../modules/UI/components/FormattedText/FormattedText.ui.js'
import { THEME } from '../../theme/variables/airbitz.js'
import { type GuiWallet } from '../../types/types.js'
import { scale, scaleH } from '../../util/scaling.js'
import { WalletListMenuModal } from '../modals/WalletListMenuModal.js'
import { Airship } from '../services/AirshipInstance.js'
import { ProgressPie } from './ProgressPie.js'

type OwnProps = {
  differencePercentage: string,
  differencePercentageStyle: any,
  exchangeRate: string,
  exchangeRateFiatSymbol: string,
  fiatBalance: string,
  fiatBalanceSymbol: string,
  cryptoAmount: string,
  guiWallet: GuiWallet,
  walletProgress: number
}

type DispatchProps = {
  selectWallet(walletId: string, currencyCode: string): void
}

type Props = OwnProps & DispatchProps

class WalletListRowComponent extends React.PureComponent<Props> {
  _onPressSelectWallet = (walletId, currencyCode, publicAddress) => {
    this.props.selectWallet(walletId, currencyCode)
    // if it's EOS then we need to see if activated, if not then it will get routed somewhere else
    // if it's not EOS then go to txList, if it's EOS and activated with publicAddress then go to txList
    const SPECIAL_CURRENCY_INFO = getSpecialCurrencyInfo(currencyCode)
    if (!SPECIAL_CURRENCY_INFO.isAccountActivationRequired || (SPECIAL_CURRENCY_INFO.isAccountActivationRequired && publicAddress)) {
      Actions[TRANSACTION_LIST]({ params: 'walletList' })
    }
  }

  openWalletListMenuModal = async () => {
    const { guiWallet } = this.props
    await Airship.show(bridge => (
      <WalletListMenuModal
        bridge={bridge}
        walletId={guiWallet.id}
        walletName={guiWallet.name}
        currencyCode={guiWallet.currencyCode}
        image={guiWallet.symbolImage}
      />
    ))
  }

  render() {
    const {
      differencePercentage,
      differencePercentageStyle,
      exchangeRate,
      exchangeRateFiatSymbol,
      fiatBalance,
      fiatBalanceSymbol,
      cryptoAmount,
      guiWallet,
      walletProgress
    } = this.props
    const { id, currencyCode, name, receiveAddress, symbolImageDarkMono } = guiWallet
    return (
      <View style={{ width: '100%' }}>
        <View>
          <TouchableHighlight
            style={styles.rowContainer}
            underlayColor={THEME.COLORS.ROW_PRESSED}
            onPress={() => this._onPressSelectWallet(id, currencyCode, receiveAddress.publicAddress)}
          >
            <View style={styles.rowContent}>
              <View style={styles.rowIconWrap}>
                {symbolImageDarkMono && <Image style={styles.rowCurrencyLogoAndroid} source={{ uri: symbolImageDarkMono }} resizeMode="cover" />}
                <View style={styles.rowCurrencyLogoAndroid}>
                  <ProgressPie size={rowCurrencyOverlaySize} color={THEME.COLORS.OPAQUE_WHITE_2} progress={walletProgress} />
                </View>
              </View>
              <View style={styles.walletDetailsContainer}>
                <View style={styles.walletDetailsRow}>
                  <T style={styles.walletDetailsRowCurrency}>{currencyCode}</T>
                  <T style={styles.walletDetailsRowValue}>{cryptoAmount}</T>
                </View>
                <View style={styles.walletDetailsRow}>
                  <T style={styles.walletDetailsRowName}>{name || s.strings.string_no_name}</T>
                  <View style={styles.walletDetailsFiatBalanceRow}>
                    <T style={styles.walletDetailsRowFiat}>{fiatBalanceSymbol}</T>
                    <T style={styles.walletDetailsRowFiat}>{fiatBalance}</T>
                  </View>
                </View>
                <View style={styles.walletDetailsRowLine} />
                <View style={styles.walletDetailsRow}>
                  <View style={styles.walletDetailsExchangeRow}>
                    <T style={styles.walletDetailsRowExchangeRate}>{exchangeRateFiatSymbol}</T>
                    <T style={styles.walletDetailsRowExchangeRate}>{exchangeRate}</T>
                  </View>
                  <T style={differencePercentageStyle}>{differencePercentage}</T>
                </View>
              </View>
              <TouchableWithoutFeedback onPress={this.openWalletListMenuModal}>
                <View style={styles.rowOptionsWrap}>
                  <Text style={styles.rowOptionsIcon}>{WALLET_LIST_OPTIONS_ICON}</Text>
                </View>
              </TouchableWithoutFeedback>
            </View>
          </TouchableHighlight>
        </View>
      </View>
    )
  }
}

const rowCurrencyOverlaySize = scale(23.3)
const rawStyles = {
  rowContainer: {
    padding: scale(6),
    paddingLeft: scale(8),
    height: scale(106),
    backgroundColor: THEME.COLORS.WHITE,
    borderBottomWidth: scale(1),
    borderBottomColor: THEME.COLORS.GRAY_3
  },
  rowContent: {
    flex: 1,
    flexDirection: 'row'
  },
  rowIconWrap: {
    alignItems: 'center',
    justifyContent: 'center',
    width: scale(36)
  },
  rowCurrencyLogoAndroid: {
    position: 'absolute',
    top: 8,
    left: 0,
    right: 0,
    bottom: 0,
    height: scale(23),
    width: scale(23),
    marginRight: scale(12),
    marginLeft: scale(3),
    resizeMode: 'contain',
    alignSelf: 'center'
  },
  rowOptionsWrap: {
    width: scaleH(37),
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center'
  },
  rowOptionsIcon: {
    fontSize: scale(20),
    color: THEME.COLORS.GRAY_1
  },
  symbol: {
    fontFamily: THEME.FONTS.SYMBOLS
  },
  walletDetailsContainer: {
    flex: 1,
    flexDirection: 'column',
    marginTop: scale(5)
  },
  walletDetailsRow: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  walletDetailsRowLine: {
    height: 1,
    borderColor: 'rgba(14, 75, 117, 0.5)',
    borderBottomWidth: 1,
    marginTop: scale(12),
    marginBottom: scale(9)
  },
  walletDetailsRowCurrency: {
    flex: 1,
    fontSize: scale(18)
  },
  walletDetailsRowValue: {
    textAlign: 'right',
    fontSize: scale(18),
    color: THEME.COLORS.GRAY_1
  },
  walletDetailsRowName: {
    flex: 1,
    fontSize: scale(14),
    color: THEME.COLORS.SECONDARY
  },
  walletDetailsRowFiat: {
    fontSize: scale(14),
    textAlign: 'right',
    color: THEME.COLORS.SECONDARY
  },
  walletDetailsRowExchangeRate: {
    fontSize: scale(14),
    textAlign: 'left',
    color: THEME.COLORS.GRAY_1
  },
  walletDetailsRowDifferenceNeutral: {
    fontSize: scale(14),
    textAlign: 'right',
    color: THEME.COLORS.SECONDARY
  },
  walletDetailsRowDifferencePositive: {
    fontSize: scale(14),
    textAlign: 'right',
    fontWeight: '400',
    color: THEME.COLORS.WALLET_LIST_DIFF_POSITIVE
  },
  walletDetailsRowDifferenceNegative: {
    fontSize: scale(14),
    textAlign: 'right',
    fontWeight: '400',
    color: THEME.COLORS.WALLET_LIST_DIFF_NEGATIVE
  },
  walletDetailsFiatBalanceRow: {
    flexDirection: 'row'
  },
  walletDetailsExchangeRow: {
    flexDirection: 'row',
    flex: 1
  }
}

export const styles: typeof rawStyles = StyleSheet.create(rawStyles)

export const WalletListRow = connect(null, (dispatch: Dispatch): DispatchProps => ({
  selectWallet(walletId: string, currencyCode) {
    dispatch(selectWallet(walletId, currencyCode, WALLET_LIST_SCENE))
  }
}))(WalletListRowComponent)
