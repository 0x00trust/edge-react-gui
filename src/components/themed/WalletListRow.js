// @flow

import * as React from 'react'
import { TouchableOpacity, View } from 'react-native'

import { memo } from '../../types/reactHooks.js'
import { type Theme, cacheStyles, useTheme } from '../services/ThemeContext.js'
import { CurrencyIcon } from './CurrencyIcon.js'
import { EdgeText } from './EdgeText.js'

type Props = {|
  children?: React.Node,
  currencyCode: string,
  exchangeRateText?: string,
  exchangeRateType?: 'neutral' | 'positive' | 'negative',
  walletName: string,

  // Icon currency:
  pluginId?: string,
  tokenId?: string,
  walletId?: string,

  // Callbacks:
  onLongPress?: () => void,
  onPress?: () => void
|}

export const WalletListRowComponent = (props: Props) => {
  const {
    children,
    currencyCode,
    walletName,

    // Icon currency:
    pluginId,
    tokenId,
    walletId,

    // Callbacks:
    onLongPress,
    onPress
  } = props
  const theme = useTheme()
  const styles = getStyles(theme)

  return (
    <TouchableOpacity style={styles.container} onLongPress={onLongPress} onPress={onPress}>
      <CurrencyIcon currencyCode={currencyCode} marginRem={1} pluginId={pluginId} sizeRem={2} tokenId={tokenId} walletId={walletId} />
      <View style={styles.detailsContainer}>
        <View style={styles.detailsTop}>
          <EdgeText style={styles.detailsCurrency}>{currencyCode}</EdgeText>
          {props.exchangeRateText != null ? (
            <EdgeText style={[styles.exchangeRateStyle, styles[props.exchangeRateType ?? 'neutral']]}>{props.exchangeRateText}</EdgeText>
          ) : null}
        </View>
        <EdgeText style={styles.detailsName}>{walletName}</EdgeText>
      </View>
      <View style={styles.childrenContainer}>{children}</View>
    </TouchableOpacity>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  // Background
  container: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'center',
    minHeight: theme.rem(4.25)
  },
  // Data containers //
  // Details Container
  detailsContainer: {
    flexDirection: 'column',
    flexGrow: 1,
    flexShrink: 1,
    marginRight: theme.rem(0.5)
  },
  // Children (Right part) Container
  childrenContainer: {
    paddingRight: theme.rem(1)
  },
  // Other styles
  detailsTop: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start'
  },
  detailsCurrency: {
    flexBasis: 'auto',
    flexShrink: 1,
    fontFamily: theme.fontFaceMedium
  },
  exchangeRateStyle: {
    textAlign: 'left',
    flexBasis: 'auto',
    flexShrink: 1,
    marginLeft: theme.rem(0.75)
  },
  detailsName: {
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  },
  // Difference Percentage Styles
  neutral: {
    color: theme.secondaryText
  },
  positive: {
    color: theme.positiveText
  },
  negative: {
    color: theme.negativeText
  }
}))

export const WalletListRow = memo(WalletListRowComponent)
