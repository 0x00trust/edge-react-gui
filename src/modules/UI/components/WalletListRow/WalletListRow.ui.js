// @flow

// import _ from 'lodash'
import { bns } from 'biggystring'
import _ from 'lodash'
import React, { Component } from 'react'
import { TouchableHighlight, View } from 'react-native'
import slowlog from 'react-native-slowlog'
import { connect } from 'react-redux'

import { intl } from '../../../../locales/intl.js'
import type { CurrencyConverter, CustomTokenInfo, GuiDenomination, GuiWallet } from '../../../../types.js'
import { getCurrencyConverter } from '../../../Core/selectors.js'
import type { Dispatch, State } from '../../../ReduxTypes'
import {
  DIVIDE_PRECISION,
  calculateWalletFiatFromCrypto,
  cutOffText,
  decimalOrZero,
  getFiatSymbol,
  getSetCurrencyMultiplier,
  getSetTokenMultiplier,
  getWalletDefaultDenomProps,
  mergeTokensRemoveInvisible,
  truncateDecimals
} from '../../../utils.js'
import { getDisplayDenomination, getExchangeDenomination } from '../../Settings/selectors.js'
import Text from '../FormattedText'
import { styles, stylesRaw } from './WalletListRowStyle.js'

export type WalletListRowOwnProps = {
  wallet: GuiWallet,
  onSelectWallet: (string, string) => void,
  excludedCurrencyCode?: string
}

export type WalletListRowStateProps = {
  fiatBalance: string,
  settings: Object,
  currencyConverter: CurrencyConverter
}

export type WalletListRowDispatchProps = {}

export type WalletListRowProps = WalletListRowOwnProps & WalletListRowStateProps & WalletListRowDispatchProps

export type WalletListRowState = {
  displayDenomination: GuiDenomination,
  exchangeDenomination: GuiDenomination,
  customTokens: Array<CustomTokenInfo>,
  fiatSymbol: string,
  isWalletFiatBalanceVisible: boolean,
  fiatBalance: string
}

export class WalletListRowComponent extends Component<WalletListRowProps, WalletListRowState> {
  constructor (props: WalletListRowProps) {
    super(props)
    slowlog(this, /.*/, global.slowlogOptions)
  }

  render () {
    const { wallet, onSelectWallet, settings, fiatBalance, excludedCurrencyCode } = this.props
    const { currencyCode, name, id, enabledTokens, nativeBalances, metaTokens } = wallet
    const denominations = wallet.allDenominations[currencyCode]
    const multiplier = getSetCurrencyMultiplier(currencyCode, settings, denominations)

    const defaultDenomProps = getWalletDefaultDenomProps(wallet, settings, currencyCode)
    const cryptoSymbol = defaultDenomProps.symbol
    const preliminaryCryptoAmount = truncateDecimals(bns.div(wallet.primaryNativeBalance, multiplier, DIVIDE_PRECISION), 6)
    const finalCryptoAmount = intl.formatNumber(decimalOrZero(preliminaryCryptoAmount, 6)) // check if infinitesimal (would display as zero), cut off trailing zeroes
    const fiatSymbol = getFiatSymbol(wallet.isoFiatCurrencyCode) || ''
    // determine tokens to render
    const enabledTokenNativeBalances = {}
    for (const currency in nativeBalances) {
      if (currency !== currencyCode && enabledTokens.includes(currency)) {
        enabledTokenNativeBalances[currency] = nativeBalances[currency]
      }
    }

    const combinedTokens = mergeTokensRemoveInvisible(metaTokens, settings.customTokens)
    const tokensToRender = []
    for (const tokenCode in enabledTokenNativeBalances) {
      if (tokenCode !== currencyCode) {
        const index = _.findIndex(combinedTokens, token => token.currencyCode === tokenCode)
        if (index >= 0) {
          tokensToRender.push(tokenCode)
        }
      }
    }
    // remove wallet options that are illogical!
    let disabled = false
    if (excludedCurrencyCode) {
      // if wallets need to be excluded
      if (currencyCode === 'ETH') {
        // if it may have tokens
        if (currencyCode === excludedCurrencyCode) {
          // if Ethereum should be disabled
          if (enabledTokens.length === 0) {
            // if ETH is excluded but has no tokens
            return null // don't show it
          } else {
            // if ETH is excluded but DOES have tokens
            disabled = true
          }
        } else {
          // if a token should be disabled
          const excludedItemIndex = tokensToRender.indexOf(excludedCurrencyCode)
          if (excludedItemIndex > -1) tokensToRender.splice(excludedItemIndex, 1)
        }
      } else {
        // if it does not have tokens
        if (currencyCode === excludedCurrencyCode) return null
      }
    }
    return (
      <View style={styles.rowWrapper}>
        <TouchableHighlight
          style={[styles.rowContainer]}
          disabled={disabled}
          underlayColor={stylesRaw.underlay.color}
          onPress={() => onSelectWallet(id, currencyCode)}
        >
          <View style={styles.rowInfo}>
            <View style={[styles.rowLeft]}>
              <Text style={[styles.rowNameText]} numberOfLines={1}>
                {cutOffText(name, 34)}
              </Text>
            </View>
            <View style={[styles.rowRight]}>
              <View style={[styles.rowRightCryptoWrap]}>
                <Text style={[styles.rowRightCryptoText]}>
                  {finalCryptoAmount} {cryptoSymbol || ''}
                </Text>
              </View>
              <View style={[styles.rowRightFiatWrap]}>
                <Text style={styles.rowRightFiatText}>
                  {fiatSymbol} {fiatBalance}
                </Text>
              </View>
            </View>
          </View>
        </TouchableHighlight>
        {tokensToRender.map(tokenCode => this.renderTokenRow(tokenCode, enabledTokenNativeBalances))}
      </View>
    )
  }

  renderTokenRow = (currencyCode: string, metaTokenBalances: Object) => {
    const { settings, wallet, currencyConverter, onSelectWallet } = this.props
    const denomination = wallet.allDenominations[currencyCode]
    const multiplier = getSetTokenMultiplier(currencyCode, settings, denomination)
    const nativeBalance = metaTokenBalances[currencyCode]
    const cryptoAmount = bns.div(nativeBalance, multiplier, DIVIDE_PRECISION)
    const parentId = wallet.id
    let fiatValue = 0
    fiatValue = currencyConverter.convertCurrency(currencyCode, wallet.isoFiatCurrencyCode, cryptoAmount)
    const fiatBalance = intl.formatNumber(fiatValue, { toFixed: 2 }) || '0'
    const fiatSymbol = getFiatSymbol(wallet.isoFiatCurrencyCode) || ''
    return (
      <TouchableHighlight
        style={[styles.tokenRowContainer]}
        underlayColor={stylesRaw.underlay.color}
        key={currencyCode}
        onPress={() => onSelectWallet(parentId, currencyCode)}
      >
        <View style={styles.currencyRowContent}>
          <View style={styles.currencyRowNameTextWrap}>
            <Text style={styles.currencyRowText}>{currencyCode}</Text>
          </View>
          <View style={[styles.rowRight]}>
            <View style={[styles.rowRightCryptoWrap]}>
              <Text style={[styles.rowRightCryptoText]}>{cryptoAmount}</Text>
            </View>
            <View style={[styles.rowRightFiatWrap]}>
              <Text style={styles.rowRightFiatText}>
                {fiatSymbol} {fiatBalance}
              </Text>
            </View>
          </View>
        </View>
      </TouchableHighlight>
    )
  }
}

const mapStateToProps = (state: State, ownProps: WalletListRowOwnProps): WalletListRowStateProps => {
  const displayDenomination = getDisplayDenomination(state, ownProps.wallet.currencyCode)
  const exchangeDenomination = getExchangeDenomination(state, ownProps.wallet.currencyCode)
  const settings = state.ui.settings
  const fiatBalance = calculateWalletFiatFromCrypto(ownProps.wallet, state)
  const currencyConverter = getCurrencyConverter(state)
  return {
    displayDenomination,
    exchangeDenomination,
    fiatBalance,
    settings,
    currencyConverter
  }
}
const mapDispatchToProps = (dispatch: Dispatch): WalletListRowDispatchProps => {
  return {}
}

export const WalletListRowConnector = connect(mapStateToProps, mapDispatchToProps)(WalletListRowComponent)
