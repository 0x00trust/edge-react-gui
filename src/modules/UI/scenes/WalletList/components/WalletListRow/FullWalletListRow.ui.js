import React, {Component} from 'react'
import strings from '../../../../../../locales/default'
import {bns} from 'biggystring'
import {
  View,
  TouchableHighlight,
  ActivityIndicator,
  Image
} from 'react-native'
import {connect} from 'react-redux'
import {Actions} from 'react-native-router-flux'
import styles from '../../style.js'
import T from '../../../../components/FormattedText'
import RowOptions from './WalletListRowOptions.ui'
import WalletListTokenRow from './WalletListTokenRowConnector.js'
import {border as b, cutOffText, truncateDecimals} from '../../../../../utils.js'
import {selectWallet} from '../../../../Wallets/action.js'
import * as SETTINGS_SELECTORS from '../../../../Settings/selectors'
import platform from '../../../../../../theme/variables/platform.js'

export const findDenominationSymbol = (denoms, value) => {
  for (const v of denoms) {
    if (v.name === value) {
      return v.symbol
    }
  }
}


class FullWalletRow extends Component {
  render () {
    return (
      <View>
        {this.props.data.item.id ? (
          <FullWalletListRowConnect data={this.props.data} />
        ) : (
          <FullListRowEmptyData />
        )}
      </View>
    )
  }
}

export default FullWalletRow

class FullWalletListRow extends Component {

  _onPressSelectWallet = (walletId, currencyCode) => {
    this.props.selectWallet(walletId, currencyCode)
    Actions.transactionList({params: 'walletList'})
  }

  render () {
    const {data} = this.props
    const walletData = data.item
    const currencyCode = walletData.currencyCode
    const cryptocurrencyName = walletData.currencyNames[currencyCode]
    const denomination = this.props.displayDenomination
    const multiplier = denomination.multiplier
    const id = walletData.id
    const name = walletData.name || strings.enUS['string_no_name']
    const symbol = denomination.symbol
    let symbolImageDarkMono = walletData.symbolImageDarkMono
    return (
      <View style={[{width: platform.deviceWidth}, b()]}>
          <View>
            <TouchableHighlight
              style={[styles.rowContainer]}
              underlayColor={'#eee'}
              {...this.props.sortHandlers}
              onPress={() => this._onPressSelectWallet(id, currencyCode)}
            >
              <View style={[styles.rowContent]}>
                <View style={[styles.rowNameTextWrap, b()]}>
                  <T style={[styles.rowNameText, b()]} numberOfLines={1}>
                  {symbolImageDarkMono
                    && <Image style={[styles.rowCurrencyLogo, b()]} transform={[{translateY: 2}]} source={{uri: symbolImageDarkMono}} resizeMode='cover' />
                  }  {cutOffText(name, 34)}</T>
                </View>
                <View style={[styles.rowBalanceTextWrap]}>
                  <T style={[styles.rowBalanceAmountText]}>
                    {truncateDecimals(bns.divf(walletData.primaryNativeBalance, multiplier).toString(), 6)}
                  </T>
                  <T style={[styles.rowBalanceDenominationText]}>{cryptocurrencyName} ({symbol || ''})</T>
                </View>
                <RowOptions sortableMode={this.props.sortableMode} executeWalletRowOption={walletData.executeWalletRowOption} walletKey={id} archived={walletData.archived} />
              </View>
            </TouchableHighlight>
            {this.renderTokenRow(id, walletData.nativeBalances, this.props.active)}
          </View>
      </View>
    )
  }

  renderTokenRow = (parentId, metaTokenBalances) => {
    let tokens = []
    for (let property in metaTokenBalances) {
      if (property !== this.props.data.item.currencyCode) {
        tokens.push(
          <WalletListTokenRow parentId={parentId}
            currencyCode={property} key={property} balance={metaTokenBalances[property]} active={this.props.active} />)
      }
    }
    return tokens
  }
}

const mapStateToProps = (state, ownProps) => {
  const displayDenomination = SETTINGS_SELECTORS.getDisplayDenomination(state, ownProps.data.item.currencyCode)
  const exchangeDenomination = SETTINGS_SELECTORS.getExchangeDenomination(state, ownProps.data.item.currencyCode)
  return {
    displayDenomination,
    exchangeDenomination
  }
}
const mapDispatchToProps = (dispatch) => ({
  selectWallet: (walletId, currencyCode) => dispatch(selectWallet(walletId, currencyCode))
})

export const FullWalletListRowConnect =  connect(mapStateToProps, mapDispatchToProps)(FullWalletListRow)

class FullListRowEmptyData extends Component {
  render () {
    return (
      <TouchableHighlight
        style={[styles.rowContainer], {height: 50, backgroundColor: 'white', padding: 16, paddingLeft: 20, paddingRight: 20, justifyContent: 'space-between', borderBottomWidth: 1, borderColor: '#EEE'}}
        underlayColor={'#eee'}
        {...this.props.sortHandlers}
      >
        <View style={[styles.rowContent]}>
          <View style={[styles.rowNameTextWrap]}>
            <ActivityIndicator style={{height: 18, width: 18}}/>
          </View>
        </View>
      </TouchableHighlight>
    )
  }
}