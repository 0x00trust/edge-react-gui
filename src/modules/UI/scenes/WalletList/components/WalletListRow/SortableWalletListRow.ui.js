import React, {Component} from 'react'
import {connect} from 'react-redux'
import strings from '../../../../../../locales/default'
import {bns} from 'biggystring'
import {
  View,
  TouchableHighlight,
  Image,
  ActivityIndicator
} from 'react-native'
import styles, {styles as styleRaw} from '../../style.js'
import T from '../../../../components/FormattedText'
import {border as b, cutOffText, truncateDecimals, findDenominationSymbol} from '../../../../../utils'
import sort from '../../../../../../assets/images/walletlist/sort.png'
import * as SETTINGS_SELECTORS from '../../../../Settings/selectors'

class SortableWalletListRow extends Component<Props, State> {

  render () {
    const {data} = this.props
    const walletData = data
    let multiplier, name, symbol, cryptoCurrencyName, symbolImageDarkMono

    // const exchangeDenomination = SETTINGS_SELECTORS.getExchangeDenomination(state, data.currencyCode)
    if (walletData.currencyCode) { // if wallet is done loading
      let displayDenomination = SETTINGS_SELECTORS.getDisplayDenominationFromSettings(this.props.settings, walletData.currencyCode)
      multiplier = displayDenomination.multiplier
      name = walletData.name || strings.enUS['string_no_name']
      symbol = findDenominationSymbol(walletData.denominations, walletData.currencyCode)
      cryptoCurrencyName = walletData.currencyNames[walletData.currencyCode]
      symbolImageDarkMono = walletData.symbolImageDarkMono
    }
    return (
      <TouchableHighlight
        style={[b('green'), styles.rowContainer, styles.sortableWalletListRow ]}
        underlayColor={styleRaw.walletRowUnderlay.color}
        {...this.props.sortHandlers}>
          {walletData.currencyCode? (
            <View style={[styles.rowContent]}>
              <View style={[styles.rowNameTextWrap]}>
                <T style={[styles.rowNameText]} numberOfLines={1}>
                  {symbolImageDarkMono
                    && <Image style={[styles.rowCurrencyLogo, b()]} transform={[{translateY: 2}]} source={{uri: symbolImageDarkMono}} resizeMode='cover' />
                  }  {cutOffText(name, 34)}
                </T>
              </View>
              <View style={[styles.rowBalanceTextWrap]}>
                <T style={[styles.rowBalanceAmountText]}>
                  {truncateDecimals(bns.div(walletData.primaryNativeBalance, multiplier, 10, 8), 6)}
                </T>
                <T style={[styles.rowBalanceDenominationText]}>{cryptoCurrencyName}
                  ({symbol || ''})</T>
              </View>
              <View style={[styles.rowDragArea, b()]}>
                <Image
                  source={sort}
                  style={{height: 15, width: 15}}
                />
              </View>
            </View>
            ) : (
              <View style={[styles.rowContent]}>
                <View style={[styles.rowNameTextWrap]}>
                  <ActivityIndicator style={{height: 18, width: 18}}/>
                </View>
              </View>
          )}
        </TouchableHighlight>
    )
  }
}

export default connect((state) => {
  const settings = state.ui.settings

  return {
    settings
  }
})(SortableWalletListRow)
