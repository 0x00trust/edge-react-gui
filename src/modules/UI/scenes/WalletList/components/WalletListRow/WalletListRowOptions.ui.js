import React, {Component} from 'react'
// import {Text, View} from 'react-native'
//import styles from '../../style'
// import Menu, {MenuOptions, MenuOption, MenuTrigger} from 'react-native-menu'
// import T from '../../../../components/FormattedText/FormattedText.ui'
import strings from '../../../../../../locales/default'
import MenuDropDown from '../../../../components/MenuDropDown/MenuDropDown.ui'
import {MenuDropDownStyle} from '../../../../../../styles/indexStyles'
export const options = [
  {
    value: 'rename',
    label: strings.enUS['string_rename']
  },{
    value: 'sort',
    label: strings.enUS['fragment_wallets_sort']
  },{
    value: 'addToken',
    label: strings.enUS['fragmet_wallets_addtoken_option']
  },/*{
    value: 'archive'
  },*/{
    value: 'delete',
    label: strings.enUS['string_delete']
  }
]

export default class WalletListRowOptions extends Component {
  constructor (props) {
    super(props)
    this.state = {
      archiveSyntax: strings.enUS['fragmet_wallets_list_'
      + (this.props.archived
        ? 'restore'
        : 'archive')
        + '_title_capitalized']
    }
  }

  optionAction = (optionKey) => {
    this.props.executeWalletRowOption(this.props.walletKey, optionKey, this.props.wallets, this.props.archives)
    if (optionKey === 'Rename') {
      this.props.updateRenameWalletInput(this.props.wallets[this.props.walletKey].name)
    }
  }

  render () {
    return (
      <MenuDropDown style={MenuDropDownStyle}
        onSelect={this.optionAction}
        data={options} />
    )
  }
}
