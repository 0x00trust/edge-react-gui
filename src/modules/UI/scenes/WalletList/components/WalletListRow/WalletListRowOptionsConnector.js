// @flow

import {connect} from 'react-redux'

import WalletListRowOptions from './WalletListRowOptions.ui'
// $FlowFixMe
import {updateRenameWalletInput} from '../../action'
// $FlowFixMe
import UI_SELECTORS from '../../../../selectors'
import type {State, Dispatch} from '../../../../../ReduxTypes.js'

const mapStateToProps = (state: State) => ({
  wallets: UI_SELECTORS.getWallets(state),
  // $FlowFixMe
  archives: state.ui.wallets.archives
})

const mapDispatchToProps = (dispatch: Dispatch) => ({
  updateRenameWalletInput: (input) => dispatch(updateRenameWalletInput(input))
})

export default connect(mapStateToProps, mapDispatchToProps)(WalletListRowOptions)
