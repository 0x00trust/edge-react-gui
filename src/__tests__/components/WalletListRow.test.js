/* globals describe it expect */
// @flow

import * as React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { WalletListRow } from '../../components/themed/WalletListRow.js'

describe('WalletListRow', () => {
  it('should render with loading props', () => {
    const renderer = new ShallowRenderer()

    const props = {
      children: 'Hello',
      currencyCode: 'BTC',
      gradient: true,
      walletName: 'My bitcoin wallet',
      onLongPress: () => undefined,
      onPress: () => undefined
    }
    const actual = renderer.render(<WalletListRow {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
