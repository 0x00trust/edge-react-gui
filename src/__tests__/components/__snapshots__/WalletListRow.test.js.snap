// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletListRow should render with loading props 1`] = `
<TouchableOpacity
  onLongPress={[Function]}
  onPress={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 96,
    }
  }
>
  <Memo(CurrencyIconComponent)
    currencyCode="BTC"
    marginRem={1}
    sizeRem={2}
  />
  <View
    style={
      Object {
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "marginRight": 11,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
        }
      }
    >
      <WithTheme(EdgeTextComponent)
        style={
          Object {
            "flexBasis": "auto",
            "flexShrink": 1,
            "fontFamily": "Quicksand-Medium",
          }
        }
      >
        BTC
      </WithTheme(EdgeTextComponent)>
    </View>
    <WithTheme(EdgeTextComponent)
      style={
        Object {
          "color": "#A4C7DF",
          "fontSize": 17,
        }
      }
    >
      My bitcoin wallet
    </WithTheme(EdgeTextComponent)>
  </View>
  <View
    style={
      Object {
        "paddingRight": 22,
      }
    }
  >
    Hello
  </View>
</TouchableOpacity>
`;
